### PR TITLE
[fix] concatenation dependend on mode

### LINKED ISSUE #1828: XQuery parsing error involving concatenation operator ||
### What is the problem

As @jesselet originally reported in https://github.com/eXist-db/eXide/issues/173, the query `<e>||</e>` raises an error:

> Cannot compile xquery: exerr:ERROR err:XPST0003: No closing end tag found for element constructor: e [at line 3, column 2]

### What did you expect

I expected the query to return the value `<e>||</e>` without raising any error.

### Describe how to reproduce or add a test

See above.

### Context information

- eXist-db version + Git Revision hash e.g. eXist-db 4.1.0-SNAPSHOT+201804122141 /  59493034b 
- Java version: Oracle JDK 1.8.0_162
- Operating system: macOS 10.13.4
- 32 or 64 bit: 64 bit
- Any custom changes in e.g. conf.xml: none

### PR DESCRIPTION
**Pull request by line-o**
## Description:

Modify the XQuery grammar to recognize `||` as concatenation operator
only when outside of attribute or element content and string constructors.

Add and extend the test suite for concatenation operator tests.
Testing for concatenation and literal `||` in different mixed modes.

Fix some minor whitespace issues.

### Reference: 

fixes #1828

### Type of tests:

xqsuite

### CODE REVIEW

File: exist-core/src/test/xquery/xquery3/concat.xql [hunk 1]
```diff
@@ -87,11 +94,66 @@ function concat:nodes() {
...
+    let $d := "d"
+    return <a b="{ $c || $d }"/>
+};
+
+(:~
+ : The two tests below cannot be commented in.
+ : They cause a parsing error and therefore must
+ : stay like this, until the underlying issue has
+ : been fixed
+ :)
```
Comment (adamretter) [inline exist-core/src/test/xquery/xquery3/concat.xql:138]: Is there an issue that has been opened that we can paste a link to here please?

**Comment by line-o** [comment]
@adamretter one of those tests was in there. I (re-)learned about the issue when I commented it in.

**Comment by line-o** [comment]
The issue in question is an old one (I had to re-open it) https://github.com/eXist-db/exist/issues/291
We could remove those tests as they are already part of a [specific testuite](test/src/xquery/literals-in-curlies.xql)

**Comment by joewiz** [comment]
@line-o Excellent! If the pending tests are already part of a different test suite, perhaps it's best to delete these, and perhaps leave a note referencing the other test suite since they're relevant to the tested functionality?

File: exist-core/src/test/xquery/xquery3/concat.xql (hunk 1 above)
Comment (joewiz) [reply exist-core/src/test/xquery/xquery3/concat.xql:139]: @adamretter In response to your suggestion, @line-o has added a link - see the commit https://github.com/eXist-db/exist/pull/3493/commits/88a7fb0c6634dea4f9475c777b95b85141c1a6f0.

Comment (adamretter) [reply exist-core/src/test/xquery/xquery3/concat.xql:139]: Yes I saw. I was just waiting for CI.

Comment (joewiz) [reply exist-core/src/test/xquery/xquery3/concat.xql:139]: Ah! Great.

**Pull request merged**